grin/test/Test/EndToEndSpec.hs

```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE DeriveFunctor, GeneralizedNewtypeDeriving, OverloadedStrings #-}
+module Test.EndToEndSpec where
+
+import Control.Monad.Reader
+import Test.Hspec
+import Test.EndToEnd
+import Test.QuickCheck hiding (Failure)
+import Data.ByteString.Char8
+import Test.Hspec.Core.Spec hiding (pending)
+import Data.Map as Map
+
+
+main :: IO ()
+main = hspec spec
+
+spec :: Spec
+spec = do
+  it "Finds failing element in the middle." $ property $
+    -- Assumption: Min passes, Mid, Max fails, min < mid, mid <= max
+    forAll (choose (0,100)) $ \mn ->
+    forAll (choose (mn+1, mn+101)) $ \mx ->
+    forAll (choose (mn+1, mx)) $ \md -> do
+      let res = testBisectM (mn,md,mx) $ bisect "some-dir" "some-result"
+      show res `shouldBe` (show $ Result "" $ Failure Nothing $ Reason $ "Test failed in pipeline step: " ++ show md)
```

Comment:
There is no Eq instance for the Result class in Hspec. I dont want to create an orphan instance and the Show instance is good enough here.

Reply:
Good enough :)